### DISCUSSION TITLE
Is it possible to add data to a TreeItem?

### DISCUSSION THREAD

**Discussion by mark-summerfield** [fltk-rs/fltk-rs#1112] [Q&A]
I want to use a tree to present some data to the user but the underlying data is much more than just the tree path. I could get by with, say, an `i32` which I could then use as an index into, say, a `HashMap<i32, MyData>`. Is this possible?

**Comment by MoAlyousef** [comment]
Currently it can't hold data. I can add a `void *` field to the C++ class and some getter and setter, the getter will probably have to be an unsafe function.

**Reply by mark-summerfield** [reply]
Adding a `void *` is clearly the most general & versatile solution. But adding an `i32` would be safe, and could be used as an index or map key to another data structure. I suppose a half-way house would be to add a `usize` which could be a pointer or an integer?

**Reply by MoAlyousef** [reply]
So turns out FLTK's Tree Item does have a user_data field and accessor methods. So I've exposed it as it is:
```rust
use fltk::{prelude::*, *};

fn main() {
    let a = app::App::default();

    let mut win = window::Window::default()
        .with_size(300, 400);

    let mut tree = tree::Tree::default().size_of_parent().center_of_parent();
    tree.set_show_root(false);

    let item = tree.add("Route A");
    item.unwrap().set_user_data(10);

    tree.add("Route A/Node 1");
    tree.add("Route A/Node 2");
    tree.add("Route B");

    let item = tree.add("Route B/Node 1");
    item.unwrap().set_user_data(5);

    tree.set_callback(|t| {
        if let Some(s) = t.first_selected_item() {
            println!("{:?}", unsafe { s.user_data::<i32>() });
        }
    });

    win.end();
    win.show();

    a.run().unwrap();
}
```

**Reply by mark-summerfield** [reply]
Thanks!
